### PR TITLE
Introduce a unit test for spa_main

### PR DESCRIPTION
**Pull request by AaronDonahue**
This new test focuses on the spa_main routine and will check the
temporal and vertical interpolation of SPA data.  It has three parts:

1. It checks that if the timestamp is at the beginning of the month
   with the same pressure profile as the SPA data the interpolated
   data matches the SPA input data for the beginning of the month.
2. It checks the same as with (1) except when the timestamp is at
   the end of the month.
3. It picks a few days into the month and checks that the interpolated
   data remains within the bounds of the actual SPA data.
   Note, it forces the vertical interpolation to handle cases when the
   target pressure profile is outside of the data pressure profile by
   constructing a target pressure profile with a surface pressure that
   is 5% bigger.

This commit also changes where the hybrid coordinate system is read for
input SPA data, rather than using the local hybrid coordinate system, which
could be different than the SPA data, it grabs the hybrid levels directly
from the SPA data file.

Finally, using padding for the SPA source data, this commit takes care of an
issue where extrapolated data was far out of reasonable bounds for SPA data.
The new approach more closely mirrors what is currently being done in the F90
implementation of the SPA interpolation.

### CODE REVIEW

File: components/scream/src/physics/spa/spa_functions_impl.hpp
Comment (bartgol) [inline components/scream/src/physics/spa/spa_functions_impl.hpp:625]: Since p_mid(end) will _never_ be 10x larger than p_src(end), hybm(end)=10 is probably way more than enough.

That said, 1e5 works.

Comment (bartgol) [reply components/scream/src/physics/spa/spa_functions_impl.hpp:625]: Actually, hold on. It's hybm that has to be large, and hyam=0, since the formula is
```
p_ref(k) = ps*hybm(k) + p0*hyam(k);
```

**Review by bartgol — CHANGES_REQUESTED** [review]
Looks like a good cleanup! I like using `_tgt` instead of `_atm`, since it makes clear that's the output.

I think there's a bug in setting the padding for hyam/hybm.

Also, the PR title suggests you added a unit test for spa, but I don't see any new test (that's fine, I just want to understand if I'm missing something).

File: components/scream/src/physics/spa/spa_functions_impl.hpp
Comment (AaronDonahue) [reply components/scream/src/physics/spa/spa_functions_impl.hpp:437]: > 

The new test is in `src/physics/spa/tests/spa_main_test.cpp`

**Comment by AaronDonahue** [comment]
> Looks like a good cleanup! I like using `_tgt` instead of `_atm`, since it makes clear that's the output.
> 
> I think there's a bug in setting the padding for hyam/hybm.
> 
> Also, the PR title suggests you added a unit test for spa, but I don't see any new test (that's fine, I just want to understand if I'm missing something).

The new test is in src/physics/spa/tests/spa_main_test.cpp

File: components/scream/src/physics/spa/spa_functions_impl.hpp
Comment (AaronDonahue) [reply components/scream/src/physics/spa/spa_functions_impl.hpp:625]: I went with hyam because no matter what p0=10000, since its a reference surface pressure.  PS can be anything.  presumably it should never be small.  But P0 is a guarantee

**Comment by bartgol** [comment]
Wow, that file showed only minor changes while I was reviewing. I must have accidentally clicked on the menu to select the commits to diff. I'll re-review.

File: components/scream/src/physics/spa/tests/CMakeLists.txt
Comment (bartgol) [inline components/scream/src/physics/spa/tests/CMakeLists.txt:17]: The operations in spa_main are all rank-local, right? Is there any feature that is tested by running with 2+ ranks? If not, I would remove the multi-rank tests.

File: components/scream/src/physics/spa/spa_functions_impl.hpp
Comment (AaronDonahue) [reply components/scream/src/physics/spa/spa_functions_impl.hpp:437]: good idea

File: components/scream/src/physics/spa/spa_functions_impl.hpp
Comment (AaronDonahue) [reply components/scream/src/physics/spa/spa_functions_impl.hpp:602]: I can switch to the inner levels if that is easier to understand.

File: components/scream/src/physics/spa/spa_functions_impl.hpp
Comment (AaronDonahue) [reply components/scream/src/physics/spa/spa_functions_impl.hpp:605]: It's hidden in the review, but if you go up to line 585 you'll see that we are looping over all of the weights.  Which makes sense, this is applying the horizontal interpolation.  So we would expect 1+ weights per point.

File: components/scream/src/physics/spa/spa_functions_impl.hpp
Comment (AaronDonahue) [reply components/scream/src/physics/spa/spa_functions_impl.hpp:625]: That being said, if switching to PS makes more sense I'm happy with that.

File: components/scream/src/physics/spa/tests/spa_main_test.cpp
Comment (bartgol) [inline components/scream/src/physics/spa/tests/spa_main_test.cpp:269]: It appears to me that you are doing both time and vertical interpolation, and checking that minmax(X_out) is within minmax of both X_beg and X_end.

I think that's a weak form test. I think you should test time and vertical interpolation separately. For time, you need a pointwise min(X_beg,X_end) < X_out < max(X_beg,X_end), which is a stronger condition than minmax over column begin bounded by minmax of beg/end. 

For vertical interpolation, instead, I would pick X_beg=X_end (for simplicity). I would also consider writing a helper function that computes min/max of a 1d view:
```
template<typename ViewT>
std::pair<Real,Real> minmax (const ViewT& v_out);
```
and loop over idof (and nbands), calling minmax on subviews of src and tgt fields, and checking their values. This might remove some code duplication, and make the intent of the test (min(in)<output<max(in) over column) more obvious to the reader.

**Review by bartgol — COMMENTED** [review]
I have a couple more comments.

File: components/scream/src/physics/spa/spa_functions_impl.hpp
Comment (bartgol) [reply components/scream/src/physics/spa/spa_functions_impl.hpp:625]: Ah, yeah, you're right, it's the same.

File: components/scream/src/physics/spa/spa_functions_impl.hpp
Comment (bartgol) [reply components/scream/src/physics/spa/spa_functions_impl.hpp:602]: Yeah, I think something like

```
for (int k=0; k<nlev; ++k)
  f_padded(k+1) = f(k);
}
f_padded(0) = 0;
f_padded(nlev) = f(nlev-1);
```
is a bit more readable.

**Comment by AaronDonahue** [comment]
I still have some fixes to do given the reviews.  So I'll work on those.  I also wanted to comment on the AT results.  The new test should not have interfered with the existing SPA tests, which appear to be failing.  I'll jump on blake today and try to narrow down what the issue is.

File: components/scream/src/physics/spa/tests/CMakeLists.txt
Comment (AaronDonahue) [reply components/scream/src/physics/spa/tests/CMakeLists.txt:17]: Good point, the reading of data should be tested on multiple ranks, but that is handled by it's own unit test, so I can delete this one running on multiple ranks.

**Comment by AaronDonahue** [comment]
I'm not sure why this keeps failing mappy for the CIME SCREAMv1 build?  I just tested on quartz and it was able to build.

**Comment by bartgol** [comment]
```
/home/e3sm-jenkins/jenkins-ws/workspace/SCREAM_PullRequest_Autotester_Mappy/2117/scream/externals/ekat/src/ekat/util/ekat_catch_main.cpp:3:10: fatal error: catch2/catch.hpp: No such file or directory
    3 | #include "catch2/catch.hpp"
      |          ^~~~~~~~~~~~~~~~~~
```

I'm gonna guess the submodule wasn't properly updated (though I have no idea why).

**Comment by bartgol** [comment]
Yup:

```
$ pwd
/home/e3sm-jenkins/jenkins-ws/workspace/SCREAM_PullRequest_Autotester_Mappy/2117/scream
$ git status
HEAD detached from 4219add
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)
  (commit or discard the untracked or modified content in submodules)

	modified:   externals/ekat (modified content)

no changes added to commit (use "git add" and/or "git commit -a")
```
@jgfouca maybe something is off with our jenkins_common scripts?

**Comment by bartgol** [comment]
Ah, yes, jenkins common is not updating submodules after merging `origin/master`. I will put up a ninja fix

**Comment by bartgol** [comment]
@AaronDonahue I suspect the following happened.

Your branch is a bit old, and the last ekat update changed ekat's submodules. When switching branches back/forth in test-all-scream, the Catch2 subdir is added/removed from the index, but it stays in the tree. This can cause issues when checking out the new ekat version, which should create the Catch2 subfodler (for the sub-submodule). Perhaps rebasing on master can help.

**Comment by AaronDonahue** [comment]
ahhh okay.  I can rebase

**Comment by AaronDonahue** [comment]
@bartgol , I will need a re-review on both of my PR's since I rebased them.

File: components/scream/src/physics/spa/tests/spa_main_test.cpp
Comment (bartgol) [inline components/scream/src/physics/spa/tests/spa_main_test.cpp:376]: Shouldn't the end point be `nlevs`? Is `spa_data_beg.nlevs=actual_nlevs+2`?

**Review by bartgol — COMMENTED** [review]
I have some comments regarding test legibility, and how to hide common lengthy (and somewhat hard to quickly parse) stuff inside some lambdas for the unit test.

I'll let you decide whether that's useful or too time consuming.

File: components/scream/src/physics/spa/tests/spa_main_test.cpp
Comment (bartgol) [inline components/scream/src/physics/spa/tests/spa_main_test.cpp:369]: Aren't you missing a similar line for all other band quantities?

Comment (AaronDonahue) [reply components/scream/src/physics/spa/tests/spa_main_test.cpp:369]: duh

Comment (AaronDonahue) [reply components/scream/src/physics/spa/tests/spa_main_test.cpp:370]: done

File: components/scream/src/physics/spa/tests/spa_main_test.cpp
Comment (bartgol) [inline components/scream/src/physics/spa/tests/spa_main_test.cpp:374]: Since you added dof_i and n in the lambda signature, you could move the definition out of the dof_i loop, so you create one lambda only once.

**Review by bartgol — APPROVED** [review]

**Comment by AaronDonahue** [comment]
This PR has been approved and the DIFFs reported on `mappy` are expected since this is non-BFB in SPA.  I'm going to manually merge.

**Pull request merged**
